File: src/oca/java/src/org/opennebula/client/user/UserPool.java
/*******************************************************************************
 * Copyright 2002-2015, OpenNebula Project (OpenNebula.org), C12G Labs
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/
package org.opennebula.client.user;

import java.util.AbstractList;
import java.util.Iterator;


import org.opennebula.client.Client;
import org.opennebula.client.OneResponse;
import org.opennebula.client.Pool;
import org.opennebula.client.PoolElement;
import org.w3c.dom.Node;

/**
 * This class represents an OpenNebula user pool.
 * It also offers static XML-RPC call wrappers.
 */
public class UserPool extends Pool implements Iterable<User>{

    private static final String ELEMENT_NAME = "USER";
    private static final String INFO_METHOD  = "userpool.info";

    /**
     * Creates a new user pool
     * @param client XML-RPC Client.
     */
    public UserPool(Client client)
    {
        super(ELEMENT_NAME, client, INFO_METHOD);
    }

    @Override
    public PoolElement factory(Node node)
    {
        return new User(node, client);
    }

    /**
     * Retrieves all the users in the pool.
     *
     * @param client XML-RPC Client.
     * @return If successful the message contains the string
     * with the information returned by OpenNebula.
     */
    public static OneResponse info(Client client)
    {
        return Pool.info(client, INFO_METHOD);
    }

    /**
     * Loads the xml representation of the user pool.
     */
    public OneResponse info()
    {
        return super.info();
    }

    public Iterator<User> iterator()
    {
        AbstractList<User> ab = new AbstractList<User>()
        {
            public int size()
            {
                return getLength();
            }

            public User get(int index)
            {
                return (User) item(index);
            }
        };

        return ab.iterator();
    }

    /**
     * Returns the User with the given Id from the pool. If it is not found,
     * then returns null. The method {@link #info()} must be called before.
     *
     * @param id of the User to retrieve
     * @return The Image with the given Id, or null if it was not found.
     */
    public User getById(int id)
    {
        return (User) super.getById(id);
    }
}

File: src/oca/ruby/opennebula/user_pool.rb
# -------------------------------------------------------------------------- #
# Copyright 2002-2024, OpenNebula Project, OpenNebula Systems                #
#                                                                            #
# Licensed under the Apache License, Version 2.0 (the "License"); you may    #
# not use this file except in compliance with the License. You may obtain    #
# a copy of the License at                                                   #
#                                                                            #
# http://www.apache.org/licenses/LICENSE-2.0                                 #
#                                                                            #
# Unless required by applicable law or agreed to in writing, software        #
# distributed under the License is distributed on an "AS IS" BASIS,          #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   #
# See the License for the specific language governing permissions and        #
# limitations under the License.                                             #
#--------------------------------------------------------------------------- #


require 'opennebula/pool'

module OpenNebula
    class UserPool < Pool
        #######################################################################
        # Constants and Class attribute accessors
        #######################################################################

        USER_POOL_METHODS = {
            :info => "userpool.info"
        }

        #######################################################################
        # Class constructor & Pool Methods
        #######################################################################

        # +client+ a Client object that represents a XML-RPC connection
        def initialize(client)
            super('USER_POOL','USER',client)
        end

        # Factory method to create User objects
        def factory(element_xml)
            OpenNebula::User.new(element_xml,@client)
        end

        #######################################################################
        # XML-RPC Methods for the User Object
        #######################################################################

        # Retrieves all the Users in the pool.
        def info()
            super(USER_POOL_METHODS[:info])
        end

        def info_all()
            return super(USER_POOL_METHODS[:info])
        end

        def info_mine()
            return super(USER_POOL_METHODS[:info])
        end

        def info_group()
            return super(USER_POOL_METHODS[:info])
        end

        alias_method :info!, :info
        alias_method :info_all!, :info_all
        alias_method :info_mine!, :info_mine
        alias_method :info_group!, :info_group

        #######################################################################
        # Helpers to get information
        #######################################################################

        # Returns a list of user IDs that have the given password. info() needs
        # to be called before.
        #
        # @param driver [String] auth driver to match
        # @param password [String] password to match
        # @return [Array] an array of IDs
        def users_with_password(driver, password)
            ids = []

            each do |user|
                if user["PASSWORD"] == password
                    ids << user["ID"]
                end
            end

            return ids
        end
    end
end
